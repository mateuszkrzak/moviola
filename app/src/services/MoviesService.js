import Axios from 'axios';

export const API_URL = 'http://react-cdp-api.herokuapp.com/movies';

async function getMovies(options) {
  const response = await Axios.get(API_URL, {
    params: {
      ...options,
      sortOrder: 'desc',
    },
  });

  return response;
}

async function getMoviesByGenres(genres) {
  const params = new URLSearchParams();
  genres.forEach(value => params.append('filter', value));

  const response = await Axios.get(API_URL, { params });

  return response;
}

async function getMovieById(id) {
  const response = await Axios.get(`${API_URL}/${id}`);

  return response;
}

export default { getMovies, getMoviesByGenres, getMovieById };
