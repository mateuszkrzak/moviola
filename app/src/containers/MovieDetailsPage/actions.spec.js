import createMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import getMovieByIdAndSimilarMoviesAsync from './actions';
import MoviesService from '../../services/MoviesService'; // eslint-disable-line

const middlewares = [thunk];
const mockStore = createMockStore(middlewares);
const mockMovie = {
  title: 'Test tile',
  release_date: '2020-10-20',
  vote_average: 9,
  poster_path: 'http://poster.path',
  runtime: 120,
  overview: 'Test overview',
  genres: ['Test genre', 'Another test genre'],
};

jest.mock('../../services/MoviesService');

describe('Details actions', () => {
  let store;

  describe('get movie by id and similar movies', () => {
    it('creates FETCH_MOVIE_BY_ID_SUCCESS and FETCH_MOVIES_BY_GENRES_SUCCESS actions when requests is successful', async () => {
      MoviesService.getMovieById.mockImplementationOnce(() => ({
        data: mockMovie,
      }));
      MoviesService.getMoviesByGenres.mockImplementationOnce(() => ({
        data: { data: [mockMovie, mockMovie], total: 2 },
      }));
      store = mockStore({
        details: {
          movie: {
            genres: ['Test genre', 'Another test genre'],
          },
        },
      });

      await store.dispatch(getMovieByIdAndSimilarMoviesAsync());

      expect(store.getActions()).toMatchSnapshot();
    });

    it('should catch errors when getMovieById request fails', async () => {
      MoviesService.getMovieById.mockImplementationOnce(() => {
        throw new Error('getMovieById error message');
      });

      store = mockStore();

      await store.dispatch(getMovieByIdAndSimilarMoviesAsync());

      expect(store.getActions()).toMatchSnapshot();
    });

    it('should catch errors when getMoviesByGenres request fails', async () => {
      MoviesService.getMovieById.mockImplementationOnce(() => ({
        data: mockMovie,
      }));
      MoviesService.getMoviesByGenres.mockImplementationOnce(() => {
        throw new Error('getMoviesByGenres error message');
      });
      store = mockStore({
        details: {
          movie: {
            genres: ['Test genre', 'Another test genre'],
          },
        },
      });

      await store.dispatch(getMovieByIdAndSimilarMoviesAsync());

      expect(store.getActions()).toMatchSnapshot();
    });
  });
});
