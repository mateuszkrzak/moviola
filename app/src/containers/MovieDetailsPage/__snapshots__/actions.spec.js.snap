// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details actions get movie by id and similar movies creates FETCH_MOVIE_BY_ID_SUCCESS and FETCH_MOVIES_BY_GENRES_SUCCESS actions when requests is successful 1`] = `
Array [
  Object {
    "type": "FETCH_MOVIE_BY_ID_REQUEST",
  },
  Object {
    "payload": Object {
      "genres": Array [
        "Test genre",
        "Another test genre",
      ],
      "overview": "Test overview",
      "poster_path": "http://poster.path",
      "release_date": "2020-10-20",
      "runtime": 120,
      "title": "Test tile",
      "vote_average": 9,
    },
    "type": "FETCH_MOVIE_BY_ID_SUCCESS",
  },
  Object {
    "type": "FETCH_MOVIES_BY_GENRES_REQUEST",
  },
  Object {
    "payload": Object {
      "data": Array [
        Object {
          "genres": Array [
            "Test genre",
            "Another test genre",
          ],
          "overview": "Test overview",
          "poster_path": "http://poster.path",
          "release_date": "2020-10-20",
          "runtime": 120,
          "title": "Test tile",
          "vote_average": 9,
        },
        Object {
          "genres": Array [
            "Test genre",
            "Another test genre",
          ],
          "overview": "Test overview",
          "poster_path": "http://poster.path",
          "release_date": "2020-10-20",
          "runtime": 120,
          "title": "Test tile",
          "vote_average": 9,
        },
      ],
      "total": 2,
    },
    "type": "FETCH_MOVIES_BY_GENRES_SUCCESS",
  },
]
`;

exports[`Details actions get movie by id and similar movies should catch errors when getMovieById request fails 1`] = `
Array [
  Object {
    "type": "FETCH_MOVIE_BY_ID_REQUEST",
  },
  Object {
    "payload": "getMovieById error message",
    "type": "FETCH_MOVIE_BY_ID_FAILURE",
  },
]
`;

exports[`Details actions get movie by id and similar movies should catch errors when getMoviesByGenres request fails 1`] = `
Array [
  Object {
    "type": "FETCH_MOVIE_BY_ID_REQUEST",
  },
  Object {
    "payload": Object {
      "genres": Array [
        "Test genre",
        "Another test genre",
      ],
      "overview": "Test overview",
      "poster_path": "http://poster.path",
      "release_date": "2020-10-20",
      "runtime": 120,
      "title": "Test tile",
      "vote_average": 9,
    },
    "type": "FETCH_MOVIE_BY_ID_SUCCESS",
  },
  Object {
    "type": "FETCH_MOVIES_BY_GENRES_REQUEST",
  },
  Object {
    "payload": "getMoviesByGenres error message",
    "type": "FETCH_MOVIES_BY_GENRES_FAILURE",
  },
]
`;
